### ISSUE TITLE
[css-position] Static position of abspos top layer elements inside fixed pos.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#9939]
This is the cyclic / ordering issue that I was concerned about in #8040.

Both Firefox and Blink get it wrong. WebKit gets it right, but I'm not sure what they're doing, maybe they're laying out the abspos twice?

```html
<!DOCTYPE html>
<meta charset="utf-8">
<style>
#outer {
  position: fixed;
  width: 200px;
  height: 200px;
  border: 2px solid blue;
}
dialog {
  display: inline;
  position: absolute;
  inset: auto;
}
</style>
Some content.
<br>
<br>
<br>
<br>
<br>
Some more.
<br>
<div id=outer>
  Fixed-pos.
  <dialog>Absolute popover inside fixed element</dialog>
</div>
<script>
  document.querySelector("dialog").showModal();
</script>
```

The main issue here is that you have:

 * Viewport box
   * ICB
     * Abspos
   * Fixedpos

So you lay out the abspos _before_ the fixedpos, and the static position isn't computed yet.

cc @rlbuis @bfgeek @chrishtr

**Comment by tabatkins** [comment]
My preferred fix for this, fwiw, is that when an element is top-layer'd, it loses its static position. (We default it to 0,0 or something equally trivial.) We have Anchor Position now, there's no need to continue twisting ourselves into knots to maintain this clunky static-pos behavior in weird cases.

This would kill the cyclic problem in all permutations of this issue.

**Comment by emilio** [comment]
@rwlbuis do you know what WebKit is doing here?

**Comment by rwlbuis** [comment]
> @rwlbuis do you know what WebKit is doing here?

Tim is more an expert here.

**Comment by nt1m** [comment]
@Emilio Not sure this answers your question, but here's where we implemented the containing block adjustments:

 https://github.com/WebKit/WebKit/commit/4f453d4c712f2989a3998bc9f604df3d95c8be20
 https://github.com/WebKit/WebKit/commit/9dffcf60597eab8cefc3792f7886abdc38b39075

**Comment by emilio** [comment]
@nt1m so webkit parents all top layer elements (abspos and fixed pos) to the viewport rather than the root scroller? Isn't that incorrect?

**Comment by chrishtr** [comment]
I agree that Chrome appears to render this example wrong (I made the example into [a jsbin here](https://output.jsbin.com/gameleyeci)). What's wrong about it is that it has an incorrect static position that includes the content under the ICB ("some content", a few `<br>`s, and "some more"). Since the containing block of the abspos element is the viewport in this case, those should not contribute to its static position.

It looks like one can see the bug more clearly in [this demo](https://output.jsbin.com/zizojefuko/1) which uses position:relative and `top:10px`, which positions correctly in Chrome. Before adding `top:10px` the static position is also wrong, but after adding `top:10px` it renders correctly in Chrome.

Perhaps there is an issue with circularity, or at least extra complications, in computing static positions correctly in cases like this, though I think it's just a straight-up bug in Chrome that I think could be fixed by computing static positions separately in the top layer. @bfgeek WDYT?

**Comment by emilio** [comment]
> I think could be fixed by computing static positions separately in the top layer.

Not sure about that. You can have this case (abspos element in the top layer depending on fixed pos not in the top layer), but you can also trivially have the opposite. Also, you can have an abspos in the top layer depend on a fixed pos in the top layer and vice versa too, right?

**Comment by chrishtr** [comment]
> You can have this case (abspos element in the top layer depending on fixed pos not in the top layer),

It doesn't depend on it though, right? Which is achieved by reparenting its box in the layout tree.

> but you can also trivially have the opposite. Also, you can have an abspos in the top layer depend on a fixed pos in the top layer and vice versa too, right?

I don't think there should be any layout dependency. The top layer should be completely independent of the ICB and its descendants.

**Comment by emilio** [comment]
Well that's what this issue is about. The dependency comes from the static position, because that is usually "the position the box would have if not abspos" (and thus if not in the top layer).

**Comment by chrishtr** [comment]
Update: looks like it actually renders correctly in Chrome. It pushes down the static position because the top layer is a sibling of the HTML element in the layout tree. [Here's an example](https://output.jsbin.com/fuzevicemo/2/
) with an artifically larger HTML element (see red rectangle).

Note: in a few times while making that demo I observed wrong static positioning for the top layer element, which seems like a race condition in Chromium. @bfgeek says he might know what caused that, but it's an implementation caching bug, not a spec issue.

**Comment by emilio** [comment]
I'm confused, so in chrome the top-layer reparenting also reparents the regular in-flow position? That is... not what I'd expect...

I mean, that makes the static position rather useless I guess, but that means that we have some leeway here and @tabatkins's proposal (making top layer elements just lose its static position) shouldn't be a webcompat issue (which is something that I feared initially).

**Comment by chrishtr** [comment]
> I'm confused, so in chrome the top-layer reparenting also reparents the regular in-flow position? That is... not what I'd expect...

What do you mean by "reparents the regular in-flow position"? Not sure what you're saying..

**Comment by emilio** [comment]
Basically that the static position doesn't reflect the position the element would have if not top layered (which is what WebKit and Gecko do). It moves the static position to after the html element.

**Comment by chrishtr** [comment]
> It moves the static position to after the html element.

Ok thanks for explaining.

That's correct, but it's what I would expect. Why would you want the static position to reflect the position it would have if not top layered? It shouldn't retain such positioning right? Because the top layer is defined to reparent and relayout in full...

**Comment by fantasai** [comment]
@emilio I put in some minimal edits to define the ICB as the static position rectangle for top layer elements. Is there anything else you think needs clarification here? https://drafts.csswg.org/css-position/#staticpos-rect

**Comment by emilio** [comment]
@fantasai looks good!
